src/app/data-client.ts

```diff
@@ -709,6 +714,147 @@ export class DataClient {
     return datasets;
   }
 
+  /**
+   * Uploads the content and metadata for tabular data.
+   *
+   * Upload tabular data collected on a robot through a specific component
+   * (e.g., a motor) along with the relevant metadata to app.viam.com. Tabular
+   * data can be found under the "Sensors" subtab of the Data tab on
+   * app.viam.com.
+   *
+   * @param tabularData The list of data to be uploaded, represented tabularly
+   *   as an array.
+   * @param partId The part ID of the component used to capture the data
+   * @param componentType The type of the component used to capture the data
+   *   (e.g., "movementSensor")
+   * @param componentName The name of the component used to capture the data
+   * @param methodName The name of the method used to capture the data.
+   * @param tags The list of tags to allow for tag-based filtering when
+   *   retrieving data
+   * @param dataRequestTimes Array of Date arrays, each containing two `Date`
+   *   objects denoting the times this data was requested[0] by the robot and
+   *   received[1] from the appropriate sensor. Passing a list of tabular data
+   *   and Timestamps with length n > 1 will result in n datapoints being
+   *   uploaded, all tied to the same metadata.
+   * @returns The file ID of the uploaded data
+   */
+  async tabularDataCaptureUpload(
+    tabularData: Record<string, googleStructPb.JavaScriptValue>[],
+    partId: string,
+    componentType: string,
+    componentName: string,
+    methodName: string,
+    tags?: string[],
+    dataRequestTimes?: Date[][]
+  ) {
+    if (dataRequestTimes?.length !== tabularData.length) {
+      throw new Error('dataRequestTimes and data lengths must be equal.');
+    }
+
+    const { dataSyncService: service } = this;
+
+    const metadata = new dataSyncPb.UploadMetadata();
+    metadata.setPartId(partId);
+    metadata.setComponentType(componentType);
+    metadata.setComponentName(componentName);
+    metadata.setMethodName(methodName);
+    metadata.setType(dataSyncPb.DataType.DATA_TYPE_TABULAR_SENSOR);
+    metadata.setTagsList(tags ?? []);
+
+    const sensorContents: dataSyncPb.SensorData[] = [];
+    for (const [i, data] of tabularData.entries()) {
+      const sensorData = new dataSyncPb.SensorData();
+
+      const sensorMetadata = new dataSyncPb.SensorMetadata();
+      const dates = dataRequestTimes[i];
+      if (dates) {
+        sensorMetadata.setTimeRequested(Timestamp.fromDate(dates[0]));
+        sensorMetadata.setTimeReceived(Timestamp.fromDate(dates[1]));
+      }
```

Comment:
I have spent an insane amount of time trying to make lint happy. Things I have tried:
- `if (dates)` "'Date | undefined' is not assignable to parameter of type 'Date'"
- `if (dates && dates[0] && dates[1])` "Prefer using an optional chain expression instead, as it's more concise and easier to read"
- `Timestamp.fromDate(dates?.[0])` "Unnecessary optional chain on a non-nullish value"
-